### PR TITLE
Persistence states

### LINKED ISSUE #141: Revise TXN state machine
Implement state transitions necessary to integrate with persistence

### PR DESCRIPTION
**Pull request by iafuture**
fixed #141

### CODE REVIEW

File: src/k2/dto/K23SI.h [hunk 1]
```diff
@@ -342,22 +347,21 @@ struct K23SITxnPushRequest {
     K2_DEF_FMT(K23SITxnPushRequest, pvid, collectionName, key, incumbentMTR, challengerMTR);
 };
 
+K2_DEF_ENUM(EndAction,
+    Abort,
+    Commit,
+    None);
```
Comment (iafuture) [inline src/k2/dto/K23SI.h:353]: None first as default

File: src/k2/module/k23si/Persistence.cpp [hunk 2]
```diff
@@ -29,54 +29,87 @@ Persistence::Persistence() {
...
+            (dto::Verbs::K23SI_Persist, request, *_remoteEndpoint, _config.persistenceTimeout());
+       })
+       .then([proms=std::move(proms)] (auto&& result) mutable {
+            auto& [status, _] = result;
+            // extract the RPC status and send it down the chain
+            K2LOG_D(log::skvsvr, "flushed with status: {}. Notifying {} promises", status, proms.size());
+            for (auto& prom: proms) {
+                prom.set_value(status);
+            }
+            return seastar::make_ready_future<Status>(std::move(status));
```
Comment (jfunston) [inline src/k2/module/k23si/Persistence.cpp:98]: I think there should be a handle_exception block here that fulfills the promises with exceptions.

File: src/k2/module/k23si/Persistence.cpp [hunk 3]
```diff
@@ -29,54 +29,87 @@ Persistence::Persistence() {
...
 }
+
+seastar::future<Status> Persistence::_chainFlushResponse() {
+    seastar::promise<Status> prom;
+    auto fut = prom.get_future();
+    _flushFut = _flushFut.then([prom = std::move(prom)](auto&& status) mutable {
+        // set status for the captured promise as well as chained futures
+        prom.set_value(status);
+        return seastar::make_ready_future<Status>(std::move(status));
+    });
```
Comment (jfunston) [inline src/k2/module/k23si/Persistence.cpp:111]: Same as above with handle_exception block

File: src/k2/module/k23si/TxnManager.cpp [hunk 4]
```diff
@@ -192,247 +199,479 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
+    }
+    // #2 if equal, pick the newer transaction
+    else if (incumbent.txnId.mtr.priority == challengerMTR.priority) {
+        auto cmpResult = incumbent.txnId.mtr.timestamp.compareCertain(challengerMTR.timestamp);
+        if (cmpResult == dto::Timestamp::LT) {
+            K2LOG_D(log::skvsvr, "incumbent {} could lose push", incumbent.txnId);
+            incumbentLostConflict = true;
+        } else if (cmpResult == dto::Timestamp::EQ) {
+            // #3 if same priority and timestamp, abort on tso ID which must be unique
+            if (incumbent.txnId.mtr.timestamp.tsoId() < challengerMTR.timestamp.tsoId()) {
```
Comment (jfunston) [inline src/k2/module/k23si/TxnManager.cpp:235]: compareCertain already considers the tsoID, so I think you can remove these branches.

File: src/k2/module/k23si/TxnManager.cpp [hunk 5]
```diff
@@ -192,247 +199,479 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
+        default:
+            K2ASSERT(log::skvsvr, false, "Invalid transaction state: {}", incumbent.state);
+    }
+}
+
+seastar::future<std::tuple<Status, dto::K23SITxnEndResponse>>
+TxnManager::endTxn(dto::K23SITxnEndRequest&& request) {
+    if (request.mtr.timestamp.compareCertain(_retentionTs) < 0) {
+        // At this point this txn has gone outside RWE. All participants will self-finalize their WIs to Aborts
+        // The TR itself will be moved to FA and deleted by the TxnManager's RWE tracking
```
Comment (jfunston) [inline src/k2/module/k23si/TxnManager.cpp:334]: Would it be better to explicitly change the txn state to force abort here?

File: src/k2/module/k23si/TxnManager.cpp [hunk 6]
```diff
@@ -192,247 +199,479 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
+            return RPCResponse(dto::K23SIStatus::OK("transaction ended"), dto::K23SITxnEndResponse{});
+        });
+}
+
+seastar::future<std::tuple<Status, dto::K23SITxnEndResponse>>
+TxnManager::_endTxnRetry(TxnRecord& rec, dto::K23SITxnEndRequest&& request) {
+    K2LOG_D(log::skvsvr, "duplicate end request {}, have= {}", request, rec);
+    if (rec.finalizeAction != request.action ||
+        rec.writeKeys != request.writeKeys) { // TODO add keyHash computed by client and compare here to verify contents
+        K2LOG_D(log::skvsvr, "invalid txn end retry request: rec={}, request={}", rec, request);
```
Comment (jfunston) [inline src/k2/module/k23si/TxnManager.cpp:376]: I think this should be a warning or error log

File: src/k2/module/k23si/TxnManager.cpp [hunk 7]
```diff
@@ -192,247 +199,479 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
+    // Upon a retry, respond after the current bgtask queue has drained (potential in-flight retries)
+    seastar::promise<std::tuple<Status, dto::K23SITxnEndResponse>> prom;
+    auto fut = prom.get_future();
+    _addBgTask(rec,
+        [prom=std::move(prom), this, txnId=rec.txnId, request=std::move(request)] () mutable {
+            // after previous bg tasks completed, see if we've finalized
+            auto it = _transactions.find(txnId);
+            if (it == _transactions.end()) {
+                // normally, we won't find a record in memory after the bg tasks have been done
+                prom.set_value(std::tuple(dto::K23SIStatus::OK("transaction retry ended"), dto::K23SITxnEndResponse{}));
```
Comment (jfunston) [inline src/k2/module/k23si/TxnManager.cpp:389]: I think we talked about this before, but is this correct? We don't actually know here if the transaction was successfully committed right?

**Review by jfunston — CHANGES_REQUESTED** [review]

File: src/k2/dto/K23SI.h [hunk 8]
```diff
@@ -342,22 +345,21 @@ struct K23SITxnPushRequest {
     K2_DEF_FMT(K23SITxnPushRequest, pvid, collectionName, key, incumbentMTR, challengerMTR);
 };
 
+K2_DEF_ENUM(EndAction,
```
Comment (iafuture) [inline src/k2/dto/K23SI.h:348]: add comments to explain this better

File: src/k2/module/k23si/Persistence.h [hunk 9]
```diff
@@ -33,28 +34,30 @@ class Persistence {
...
 
     // flush all pending writes to persistence.
     seastar::future<Status> flush();
 
     // Appends are always asynchronous (buffered locally) until an explicit call to flush()
-    // The returned future is satisfied when the data is successfully persisted.
+    // append_cont returns the status of the flush call
     template<typename ValueType>
-    seastar::future<> append_cont(const ValueType& val) {
+    seastar::future<Status> append_cont(const ValueType& val) {
```
[the remark this answers is not in the document]
Comment (iafuture) [reply src/k2/module/k23si/Persistence.h:49]: done

File: src/k2/module/k23si/TxnManager.cpp (hunk 7 above)
Comment (iafuture) [reply src/k2/module/k23si/TxnManager.cpp]: you're right. Fixed to just return a KEY_NOT_FOUND for now with a comment for future optimization

File: src/k2/module/k23si/Persistence.cpp (hunk 2 above)
Comment (iafuture) [reply src/k2/module/k23si/Persistence.cpp:98]: Added then_wrapped to handle exceptions

**Review by jfunston — APPROVED** [review]

File: src/k2/module/k23si/TxnManager.cpp [hunk 10]
```diff
@@ -192,247 +199,482 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
+    auto timeout = (10s + _config.writeTimeout() * rec.writeKeys.size()) / _config.finalizeBatchSize();
+
+    if (rec.syncFinalize) {
+        return _finalizeTransaction(rec, FastDeadline(timeout));
+    }
+
+    // we're doing async finalize. enqueue in background tasks
+    _addBgTask(rec,
+        [this, &rec, timeout] {
+            return seastar::sleep(rec.timeToFinalize)
```
Comment (jerryhfeng) [inline src/k2/module/k23si/TxnManager.cpp:637]: Why adding a sleep here?

File: src/k2/module/k23si/TxnManager.cpp [hunk 11]
```diff
@@ -192,247 +199,482 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
 
-seastar::future<> TxnManager::_deleted(TxnRecord& rec) {
-    K2LOG_D(log::skvsvr, "Setting status to deleted for {}", rec);
+seastar::future<Status> TxnManager::_finalizedPIP(TxnRecord& rec) {
+    K2LOG_D(log::skvsvr, "Setting status to FinalizedPIP for {}", rec);
     // set state
-    rec.state = dto::TxnRecordState::Deleted;
+    rec.state = dto::TxnRecordState::FinalizedPIP;
     // manage hb expiry
     rec.unlinkHB(_hblist);
```
Comment (jerryhfeng) [inline src/k2/module/k23si/TxnManager.cpp:651]: These two unlink looks like redundant. Maybe we can assert here that this tr doesn't exist in these two lists for debugging purpose.

File: src/k2/module/k23si/TxnManager.cpp [hunk 12]
```diff
@@ -192,247 +199,482 @@ TxnRecord& TxnManager::_createRecord(dto::TxnId txnId) {
...
-    // set state: no change
-    // manage hb expiry
-    rec.unlinkHB(_hblist);
-    rec.hbExpiry = CachedSteadyClock::now() + 2*_hbDeadline;
-    _hblist.push_back(rec);
-    // manage rw expiry: no change
-    // persist if needed: no need
-    return seastar::make_ready_future();
+                    K2LOG_D(log::skvsvr, "Erasing txn record: {}", rec);
+                    rec.unlinkRW(_rwlist);
```
Comment (jerryhfeng) [inline src/k2/module/k23si/TxnManager.cpp:668]: when should we remove tr from RWlink?

**Review by jerryhfeng — DISMISSED** [review]

File: src/k2/module/k23si/TxnManager.cpp (hunk 10 above)
Comment (iafuture) [reply src/k2/module/k23si/TxnManager.cpp:637]: fixed and added comment

File: src/k2/module/k23si/TxnManager.cpp (hunk 11 above)
Comment (iafuture) [reply src/k2/module/k23si/TxnManager.cpp:651]: cleaned up

**Review by jerryhfeng — APPROVED** [review]

**Pull request merged**